CRM/CiviGeometry/BAO/Geometry.php

```diff
@@ -433,39 +433,147 @@ public static function wkt2kml($wkt) {
-      ->where("geo_code_1 <= '#top_bound'", ['top_bound' => $bounds['top_bound']])
-      ->where("geo_code_1 >= '#bottom_bound'", ['bottom_bound' => $bounds['bottom_bound']]);
-    $addressResults = CRM_Core_DAO::executeQuery($select->toSQL())->fetchAll();
-    $results = [];
-    if (!empty($addressResults)) {
-      $address_ids = CRM_Utils_Array::collect('id', $addressResults);
-      foreach ($address_ids as $address_id) {
-        $select = CRM_Utils_SQL_Select::from(self::getTableName() . ' g, civicrm_address ca')
-          ->select("ca.id as address_id, g.id as geometry_id")
-          ->where("ST_Contains(g.geometry, ST_GeomFromText(CONCAT('POINT(', ca.geo_code_2, ' ', ca.geo_code_1, ')'), 4326)) = 1")
-          ->where("g.id = #geometry_id", ['geometry_id' => $geometry_id])
-          ->where("ca.id = #address_id", ['address_id' => $address_id]);
-        $result = CRM_Core_DAO::executeQuery($select->toSQL())->fetchAll();
-        if (!empty($result)) {
-          $results[] = $result[0];
+    $tmpTbl = CRM_Utils_SQL_TempTable::build();
+    if ($params['keep_temp_table']) {
+      $tmpTbl->setDurable();
+    } else {
+      $tmpTbl->setAutodrop(true);
+    }
+    $tmpTbl->createWithColumns('address_id INT');
+
+    $tempTableName = $tmpTbl->getName();
+    if ($params['precheck_relationships']) {
+      CRM_Core_DAO::executeQuery("
+        INSERT INTO $tempTableName (address_id)
+        SELECT ca.id
+        FROM
+          civicrm_address ca
+        WHERE
+              ca.geo_code_2 >= %1
+          AND ca.geo_code_2 <= %2
+          AND ca.geo_code_1 <= %3
+          AND ca.geo_code_1 >= %4
+          AND NOT EXISTS (
+            SELECT *
+            FROM civigeometry_geometry_entity cge
+            WHERE
+              cge.entity_id = ca.id
+              AND cge.geometry_id = %5
+              AND cge.entity_table = 'civicrm_address'
+          )
+        ORDER BY ca.id
+      ", [
+        1 => [$bounds['left_bound'], 'Float'],
+        2 => [$bounds['right_bound'], 'Float'],
+        3 => [$bounds['top_bound'], 'Float'],
+        4 => [$bounds['bottom_bound'], 'Float'],
+        5 => [$geometry_id, 'Integer']
+      ]);
+    } else {
+      CRM_Core_DAO::executeQuery("
+        INSERT INTO $tempTableName (address_id)
+        SELECT ca.id
+        FROM civicrm_address ca
+        WHERE
+              ca.geo_code_2 >= %1
+          AND ca.geo_code_2 <= %2
+          AND ca.geo_code_1 <= %3
+          AND ca.geo_code_1 >= %4
+        ORDER BY ca.id
+      ", [
+        1 => [$bounds['left_bound'], 'Float'],
+        2 => [$bounds['right_bound'], 'Float'],
+        3 => [$bounds['top_bound'], 'Float'],
+        4 => [$bounds['bottom_bound'], 'Float'],
+      ]);
+    }
+
+    $numCandidates = CRM_Core_DAO::singleValueQuery("SELECT COUNT(*) FROM $tempTableName");
+
+    function fetchCandidateBatch($candidateTable, $offset, $batchSize) {
```

Comment:
if this function is a static and b only used in this class lets make it a private static function IMO

Reply:
Ok, I'll refactor that.